### ISSUE TITLE
Add Android support.

### ISSUE THREAD

**Issue by zdabing** [alexmercerind/dart_vlc#157]
can support Android and ios?

**Comment by alexmercerind** [comment]
iOS support is already added by contributor. Only Android is left, will appreciate any PRs.

**Comment by alexmercerind** [comment]
mpv doesn't have as friendly api as vlc. and a lot lot lot lot of work has already been done now. we share a lot of code unlike other plugins. its all native c++.

**Comment by zdabing** [comment]
> mpv doesn't have as friendly api as vlc. and a lot lot lot lot of work has already been done now. we share a lot of code unlike other plugins. its all native c++.

expect Android

**Comment by krjw-eyev** [comment]
do you expect Android support or shall we expect a pr from you with Android support?

**Comment by zdabing** [comment]
> do you expect Android support to be implemented by us or shall we expect a pr from you with Android support?

expect Android support to be implemented by you

**Comment by alexmercerind** [comment]
@krjw-eyev Could you mail me your Google mail at saini123hitesh@gmail.com. I will add you as a publisher in pub.dev to publish the updated version. I'm kinda busy with something else these days.

**Comment by prateekmedia** [comment]
@alexmercerind @krjw-eyev Why was this issue closed? This plugin doesn't support android yet.

**Comment by alexmercerind** [comment]
> @alexmercerind @krjw-eyev Why was this issue closed? This plugin doesn't support android yet.

Yeah. We aren't still decided, as the whole "plugin" uses FFI and uses common C++ shared library to support all platforms. But, on Android it seems I have to get my hands dirty with Java instead.

Although we will support every Flutter platform after that, scope of improvement for using texture API will be there. On iOS, we are not using texture, and its bad, same on Linux (Flutter team isn't reviewing PR), only Windows uses texture ATM & Android (fortunately someday I'll work). MacOS texture support is stuck at #122 (Someone with macOS/iOS, take a look & merge that without further improvements), same can be used on iOS I suppose.

**Comment by prateekmedia** [comment]
@alexmercerind I found a similar package which uses libvlc that supports android, https://github.com/arcticfox1919/vlc-flutter/tree/main/android/src/main/java/xyz/bczl/vlc_flutter

**Comment by mitchross** [comment]
I worked heavily on https://github.com/solid-software/flutter_vlc_player

**Comment by alexmercerind** [comment]
@mitchross

> I worked heavily on https://github.com/solid-software/flutter_vlc_player

Hello @mitchross, thanks for joining the conversation. [`flutter_vlc_player`](https://github.com/solid-software/flutter_vlc_player) looks nice. I recall @ mfkl from videoLAN linked that after I posted [`dart_vlc`](https://github.com/alexmercerind/dart_vlc) on Twitter. I honestly had no idea about existing libvlc project for Flutter, either way I wanted audio playback for desktop back then. Following @ mfkl's reply, I joined the Discord for libvlc & I'm pretty sure I've seen you & you've seen me there.

My work has always been specific for Windows/Linux (C++) & since there were no video playback available libraries (still not available either for Linux or Windows), I decided to work on this. From my observations, libvlc 3.0.0's video frame callbacks aren't really performant. There are some people complaining about CPU usage on Windows, I wonder how well it'd be running on Android.
Following the performance issue #184, I started working on [`flutter_native_view`](https://github.com/alexmercerind/flutter_native_view) (currently just supporting Windows) because Flutter team couldn't make platform views work on Windows.

Current implementation of [`dart_vlc`](https://github.com/alexmercerind/dart_vlc) uses `dart:ffi` & `package:ffi` for communicating with libvlc & uses platform channel interface _*just*_ for texture registration (which I will possibly remove if I start using [`flutter_native_view`](https://github.com/alexmercerind/flutter_native_view)). And I fear mixing the two separate implementation (one based on platform interface i.e. `flutter_vlc_player` & another on ffi i.e. [`dart_vlc`](https://github.com/alexmercerind/dart_vlc).

I'm not really able to spare much time for this project these days [see commit history] & I don't know if I will be able to contribute back to [`flutter_vlc_player`](https://github.com/solid-software/flutter_vlc_player) as of now. On the other hand, I think there are many good alternatives available to libvlc (in terms of _features_, _bundle-size_ or even _performance_) such as official `winrt::` media playback APIs on Windows, ExoPlayer on Android, GStreamer on Linux etc. etc. This plugin ([`dart_vlc`](https://github.com/alexmercerind/dart_vlc)) is now just "restricted" to use libvlc because of it's name. 

No fear to say, I wonder if people opened/upvoted this issue only because they didn't wanna write separate code for Android/iOS (using `video_player` or `better_player` which also allow caching & use ExoPlayer). To me, using `libvlc` on Android is bit redundant (atleast just for playing media).
There are some really really fine OSS apps on _just_ GitHub using [`dart_vlc`](https://github.com/alexmercerind/dart_vlc). From my observations, they're using it just because there are no alternatives to [`dart_vlc`](https://github.com/alexmercerind/dart_vlc). libvlc isn't really business friendly [LGPL/GPL] or good to bundle (#220). I wonder most projects will readily switch to something else once things are bit sorted out for desktop too.

Lastly, I really really appreciate you taking the initiative for the collaboration. I'm currently not really ready to drop my copyleft LGPL on this project. Still, feel free to depend upon [`dart_vlc`](https://github.com/alexmercerind/dart_vlc) as a `pub.dev` dependency in your project or projects.

@zdabing 
> can you use libmpv make dart_mpv. In my opinion 

I already have a private `libmpv` library used in [Harmonoid](https://github.com/harmonoid). An early-POC & public version is available here under permissive MIT license: [libmpv.dart](https://github.com/alexmercerind/libmpv.dart). The version used in [Harmonoid](https://github.com/harmonoid) is not available under open-source license as of now (capable of YouTube support, tag parsing & video playback).

TL;DR

I'm not going to implement Android support in [`dart_vlc`](https://github.com/alexmercerind/dart_vlc) (atleast anytime soon), because I have no revenue out of this, boss shouting at me to do this or any paycheck for this.
However, I'll be happy to review/improve any pull requests that one makes here for Android support.

Thanks.
I think I've been little too honest with the reply, regardless have a great day.

**Comment by prateekmedia** [comment]
> ...such as official `winrt::` media playback APIs on Windows, ExoPlayer on Android, GStreamer on Linux etc. etc. This plugin ([`dart_vlc`](https://github.com/alexmercerind/dart_vlc)) is now just "restricted" to use libvlc because of it's name.

The idea sounds cool, I wonder if anyone is working on something like this and calling it a `cross_platform_player`, no restriction at all.

**Comment by alexmercerind** [comment]
> The idea sounds cool, I wonder if someone is working on it and calling it a `cross_platform_player`, no restriction at all.

I also want the same. A universal media playback library for Flutter independent of internal library, using the best available for the platform.
I'd love to but I can never work on macOS/iOS support. May plan to fork `video_player` or something. 😄

**Comment by mitchross** [comment]
Hey @alexmercerind  when I dropped the comment, I saw you had a discussion board and moved more commentary to https://github.com/alexmercerind/dart_vlc/discussions/233

**Comment by alexmercerind** [comment]
Raised #254 to target platform-support centrally.
